### PR TITLE
Move GetIUnknownForObject calls to ComHelpers

### PR DESCRIPTION
**Pull request by JeremyKuhne**
We want to keep our Marshal calls in ComHelpers so we can do the right logic regarding ComWrappers vs. Marshal.

Remove AdviseHelper and use the new functionality to set up the notifications (much simpler).

Move IRichEditOleCallback to CsWin32. Add new static helper to DataObject to create from a native IDataObject.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8498)

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
There are still some in tests. There are follow up PRs in the works.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ActiveXImpl.cs [hunk 1]
```diff
@@ -1411,15 +1408,19 @@ internal unsafe HRESULT QuickActivate(QACONTAINER* pQaContainer, QACONTROL* pQaC
...
-                    {
-                        // For the default source interface, call IConnectionPoint.Advise with the supplied event sink.
-                        // This is easier said than done. See notes in AdviseHelper.AdviseConnectionPoint.
-                        AdviseHelper.AdviseConnectionPoint(_control, pQaContainer->pUnkEventSink, eventInterface, out pQaControl->dwEventCookie);
-                    }
-                    catch (Exception e) when (!ClientUtils.IsCriticalException(e))
+                    // Control doesn't explicitly implement IConnectionPointContainer, but it is generated with a CCW by
+                    // COM interop.
+
+                    using var container = ComHelpers.GetComScope<IConnectionPointContainer>(_control);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.ActiveXImpl.cs:1414]: @AaronRobinsonMSFT, @jkoritzinsky Wanted to point this out to you as a place we're using IConnectionPointContainer. I'll have to figure out how to replicate manually...

Comment (jkoritzinsky) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.ActiveXImpl.cs:1414]: The managed implementation for COM events in dotnet/runtime is located in the ComEvents files in this folder: https://github.com/dotnet/runtime/tree/main/src/libraries/Common/src/System/Runtime/InteropServices

That should hopefully help.

File: src/System.Windows.Forms/src/System/Windows/Forms/RichTextBox.OleCallback.cs [hunk 2]
```diff
@@ -6,330 +6,335 @@
 using Windows.Win32.System.Com;
 using Windows.Win32.System.Com.StructuredStorage;
 using Windows.Win32.System.Ole;
+using Windows.Win32.System.SystemServices;
+using Windows.Win32.UI.Controls.RichEdit;
+using Com = Windows.Win32.System.Com;
 using static Interop;
-using static Interop.Richedit;
-using IComDataObject = System.Runtime.InteropServices.ComTypes.IDataObject;
+//using static Interop.Richedit;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/RichTextBox.OleCallback.cs:13]: Please remove this comment

**Review by Tanya-Solyanik — APPROVED** [review]
Added some style comments

**Comment by dreddy-work** [comment]
>             else

nit:   `else` isn't needed as `if` block returns. 

---
Refers to: src/System.Windows.Forms.Primitives/src/Windows/Win32/ComHelpers.cs:129 in 06f4bcc. [](commit_id = 06f4bcc489f2f7be101058606b0177fa5a4c8ee8, deletion_comment = False)

**Review by dreddy-work — APPROVED** [review]
:shipit:

**Review by Tanya-Solyanik — APPROVED** [review]
Looks good, but consider Devendar's point.

**Pull request merged**